crates/arrow_tools/src/record_batch.rs

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use arrow::{
+    array::{
+        new_null_array, Array, Int32Array, Int64Array, LargeStringArray, RecordBatch, StringArray,
+        TimestampMicrosecondArray, TimestampMillisecondArray, TimestampSecondArray,
+    },
+    datatypes::{DataType, Field, SchemaRef, TimeUnit},
+};
+use snafu::prelude::*;
+use std::{num::TryFromIntError, sync::Arc};
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Error converting record batch: {}", source))]
+    ConvertRecordBatch { source: arrow::error::ArrowError },
+
+    #[snafu(display("Unable to convert array: {}", source))]
+    ParseTimestampFromString { source: chrono::ParseError },
+
+    #[snafu(display("Unable to convert {} array from {:?} to {:?}", name, from, to))]
+    ConvertArrowArray {
+        from: DataType,
+        to: DataType,
+        name: String,
+    },
+
+    #[snafu(display("Unable to try from int: {}", source))]
+    ConvertInt { source: TryFromIntError },
+
+    #[snafu(display("Unable to use null for field {field}"))]
+    UnableToUseNullForField { field: String },
+
+    #[snafu(display("Unable to downcast array {field} to {downcast_to}"))]
+    UnableToDowncast { field: String, downcast_to: String },
+}
+
+/// Cast a given record batch into a new record batch with the given schema.
+///
+/// # Errors
+///
+/// This function will return an error if the record batch cannot be casted.
+#[allow(clippy::needless_pass_by_value)]
+pub fn try_cast_to(record_batch: RecordBatch, schema: SchemaRef) -> Result<RecordBatch> {
+    let existing_schema = record_batch.schema();
+
+    if schema.contains(&existing_schema) {
+        return record_batch
+            .with_schema(schema)
+            .context(ConvertRecordBatchSnafu);
+    }
+
+    let cols = schema
+        .fields()
+        .into_iter()
+        .map(|field| {
+            if let (Ok(existing_field), Some(column)) = (
+                record_batch.schema().field_with_name(field.name()),
+                record_batch.column_by_name(field.name()),
+            ) {
+                if field.contains(existing_field) {
+                    Ok(Arc::clone(column))
+                } else {
+                    try_cast_array(Arc::clone(column), existing_field, field)
+                }
+            } else if field.is_nullable() {
```

Comment:
What are the scenarios in which we want to do this vs just erroring? This basically means the requested schema has a column that doesn't exist in the source schema - which I would expect to be an error case.

Reply:
Technically you can cast a record batch from less columns to more columns if the  new fields are nullable. Which can support schema expansion.